subarray.go

```diff
@@ -0,0 +1,263 @@
+func (sa *Subarray) Free() {
+	if sa.subarray != nil {
+		C.tiledb_subarray_free(&sa.subarray)
+	}
+}
+
+// SetCoalesceRanges sets coalesce_ranges property on a TileDB subarray object.
+// Intended to be used just after array.NewSubarray to replace the initial coalesce_ranges == true with coalesce_ranges = false if needed.
+func (sa *Subarray) SetCoalesceRanges(b bool) error {
+	var coalesce C.int
+	if b {
+		coalesce = 1
+	}
+
+	ret := C.tiledb_subarray_set_coalesce_ranges(sa.context.tiledbContext, sa.subarray, coalesce)
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error setting coalesce ranges on subarray: %s", sa.context.LastError())
+	}
+
+	return nil
+}
+
+// AddRange adds a range along a subarray dimension. It checks the types of range and dimension and
+// if the datatype of the range is not the same as the type of the dimension it returns an error.
+func (sa *Subarray) AddRange(dimIdx uint32, r Range) error {
+	dt, isVar, err := datatypeOfDimensionFromIndex(sa.array, dimIdx)
+	if err != nil {
+		return err
+	}
+	if err := r.assertCompatibility(dt, isVar); err != nil {
+		return err
+	}
+
+	var ret C.int32_t
+	if isVar {
+		startSlice := []byte(r.start.(string))
+		endSlice := []byte(r.end.(string))
+		ret = C.tiledb_subarray_add_range_var(sa.context.tiledbContext, sa.subarray, C.uint32_t(dimIdx),
+			unsafe.Pointer(&startSlice[0]), C.uint64_t(len(startSlice)), unsafe.Pointer(&endSlice[0]), C.uint64_t(len(endSlice)))
+	} else {
+		sp := reflect.New(reflect.ValueOf(r.start).Type())
+		sp.Elem().Set(reflect.ValueOf(r.start))
+		ep := reflect.New(reflect.ValueOf(r.end).Type())
+		ep.Elem().Set(reflect.ValueOf(r.end))
+		ret = C.tiledb_subarray_add_range(sa.context.tiledbContext, sa.subarray, C.uint32_t(dimIdx),
+			sp.UnsafePointer(), ep.UnsafePointer(), nil)
+	}
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error adding subarray range: %s", sa.context.LastError())
+	}
+
+	return nil
+}
+
+// AddRangeByName adds a range along a subarray dimension. It checks the types of range and dimension and
+// if the datatype of the range is not the same as the type of the dimension it returns an error.
+func (sa *Subarray) AddRangeByName(dimName string, r Range) error {
+	dt, isVar, err := datatypeOfDimensionFromName(sa.array, dimName)
+	if err != nil {
+		return err
+	}
+	if err := r.assertCompatibility(dt, isVar); err != nil {
+		return err
+	}
+
+	cDimName := C.CString(dimName)
+	defer C.free(unsafe.Pointer(cDimName))
+
+	var ret C.int32_t
+	if isVar {
+		startSlice := []byte(r.start.(string))
+		endSlice := []byte(r.end.(string))
+		ret = C.tiledb_subarray_add_range_var_by_name(sa.context.tiledbContext, sa.subarray, cDimName,
+			unsafe.Pointer(&startSlice[0]), C.uint64_t(len(startSlice)), unsafe.Pointer(&endSlice[0]), C.uint64_t(len(endSlice)))
+	} else {
+		sp := reflect.New(reflect.ValueOf(r.start).Type())
+		sp.Elem().Set(reflect.ValueOf(r.start))
+		ep := reflect.New(reflect.ValueOf(r.end).Type())
+		ep.Elem().Set(reflect.ValueOf(r.end))
```

Comment:
Apologies if I am missing something here, but could this be expressed as:

```go
sp := reflect.ValueOf(r.start).Addr()
// or maybe ....Elem().Addr() ?

Reply:
This was my first try but i got
```
panic: reflect.Value.Addr of unaddressable value [recovered]
	panic: reflect.Value.Addr of unaddressable value
```

Looks like reflection cannot give you the address of a value, the equivalent of `var i int; ip := &i` The only way to make it work what this, create a pointer, copy the value and get back the pointer

Reply:
Maybe pulling this out into a function would make it easier to understand, something like:

```go
// addressableValue copies the value of `val` into an addressable location.
func addressableValue(val any) reflect.Value {
  valVal := reflect.ValueOf(val)
  pointable := reflect.New(valVal.Type())
  pointable.Elem().Set(valVal)
  return pointable
}
```